File: lma-audio-capture-app-stack/source/windows/Engine/VideoCapture.cs
using System.IO;
using ScreenRecorderLib;

namespace LMA;

/// <summary>
/// Optional desktop-video capture (Windows counterpart to macOS VideoCapture):
/// records the chosen display or window at a low frame rate, H.264-encoded in
/// fragmented-MP4 (CMAF-style) via ScreenRecorderLib (which wraps
/// Windows.Graphics.Capture + Media Foundation), and hands each fMP4 chunk to a
/// callback for streaming over the video websocket.
///
/// ScreenRecorderLib writes to a Stream in fragmented mode; we plug in a
/// <see cref="CallbackStream"/> so every write becomes a socket send — nothing
/// is buffered to disk on the client.
///
/// Encoder invariants the server/pipeline relies on (matching the macOS client):
///   • Fragmented MP4 (fragmented moov + moof/mdat) so the server file is
///     playable/probe-able as it grows and ffmpeg can mux at call end.
///   • Keyframe at least every ~2 seconds (KeyFrameIntervalInSeconds) for cheap
///     seeking by future frame-extraction features.
///   • H.264, ~5 fps, capped resolution — screen content compresses well.
///
/// Video is best-effort: any failure here is logged and reported via OnFailed;
/// it never touches the audio path.
/// </summary>
public sealed class VideoCapture
{
    /// <summary>A capturable video source (display or window) for the picker.</summary>
    public readonly record struct Source(string Id, string Name);

    /// <summary>Frames per second — matches the macOS client and the VP.</summary>
    public const int Fps = 5;
    /// <summary>Cap the longest edge; enough for legible slide text.</summary>
    public const int MaxDimension = 1920;
    /// <summary>Target H.264 bitrate (bits/sec). Screen content at 5 fps is light.</summary>
    public const int BitsPerSecond = 1_500_000;

    private readonly string _sourceId;
    private readonly Action<byte[]> _onSegment;
    private Recorder? _recorder;
    private CallbackStream? _stream;
    private volatile bool _stopped;

    /// <summary>Fired (once) if capture fails to start or dies mid-recording.</summary>
    public Action<string>? OnFailed;
    /// <summary>Fired when the first frame is written (offset reporting).</summary>
    public Action? OnFirstFrame;
    private bool _firstFrameSeen;
    public DateTime? FirstFrameDate { get; private set; }

    public VideoCapture(string sourceId, Action<byte[]> onSegment)
    {
        _sourceId = sourceId;
        _onSegment = onSegment;
    }

    // MARK: - Source enumeration (Settings picker)

    /// <summary>
    /// List capturable sources beyond the default: the default "Entire screen"
    /// is the picker's built-in empty-id option; here we add any additional
    /// displays and titled windows (untitled/system windows are noise).
    /// </summary>
    public static List<Source> ListSources()
    {
        var outList = new List<Source>();
        try
        {
            var displays = Recorder.GetDisplays();
            for (int i = 1; i < displays.Count; i++) // index 0 == default "Entire screen"
                outList.Add(new Source($"display:{displays[i].DeviceName}", $"Display {i + 1}"));

            foreach (var w in Recorder.GetWindows())
            {
                if (string.IsNullOrWhiteSpace(w.Title)) continue;
                outList.Add(new Source($"window:{w.Handle}", w.Title));
            }
        }
        catch (Exception e)
        {
            Console.Error.WriteLine($"video source enumeration failed: {e.Message}");
        }
        return outList;
    }

    // MARK: - Lifecycle

    public void Start()
    {
        // Resolve the persisted source id to a ScreenRecorderLib source; fall
        // back to the primary display when it's gone (window closed, display
        // unplugged, stale id after reboot).
        RecordingSourceBase source = ResolveSource(_sourceId);

        var opts = new RecorderOptions
        {
            SourceOptions = new SourceOptions { RecordingSources = { source } },
            OutputOptions = new OutputOptions
            {
                RecorderMode = RecorderMode.Video,
                // Cap the longest edge; keep aspect ratio (0 = auto for the
                // other dimension). ScreenRecorderLib letterboxes as needed.
                OutputFrameSize = new ScreenSize(MaxDimension, 0),
                Stretch = StretchMode.Uniform,
            },
            VideoEncoderOptions = new VideoEncoderOptions
            {
                Encoder = new H264VideoEncoder
                {
                    BitrateMode = H264BitrateControlMode.UnconstrainedVBR,
                },
                Bitrate = BitsPerSecond,
                Framerate = Fps,
                // ~2s keyframe cadence => cheap seeking for frame extraction.
                Quality = 70,
                IsFragmentedMp4Enabled = true,
                IsHardwareEncodingEnabled = true,
                IsLowLatencyEnabled = true,
            },
            AudioOptions = new AudioOptions { IsAudioEnabled = false }, // audio is AudioCapture's job
            MouseOptions = new MouseOptions { IsMousePointerEnabled = true },
        };

        try
        {
            _stream = new CallbackStream(OnChunk);
            _recorder = Recorder.CreateRecorder(opts);
            _recorder.OnRecordingComplete += (_, _) => { };
            _recorder.OnRecordingFailed += (_, e) =>
            {
                if (!_stopped) OnFailed?.Invoke(e.Error);
            };
            _recorder.Record(_stream);
            Console.WriteLine($"✓ Screen video capture started (H.264 @ {Fps} fps → fMP4)");
        }
        catch (Exception e)
        {
            OnFailed?.Invoke(e.Message);
            throw;
        }
    }

    /// <summary>
    /// Mirror the audio mixer's pause state. ScreenRecorderLib's Pause stops
    /// emitting frames and resumes the timeline where it left off, so the video
    /// stays aligned with the audio recording (which also excises paused time).
    /// </summary>
    public void SetPaused(bool paused)
    {
        var rec = _recorder;
        if (rec == null) return;
        try
        {
            if (paused) rec.Pause();
            else rec.Resume();
        }
        catch (Exception e)
        {
            Console.Error.WriteLine($"video pause/resume failed: {e.Message}");
        }
    }

    /// <summary>Stop capture and flush the final fMP4 fragment(s).</summary>
    public async Task StopAsync()
    {
        _stopped = true;
        var rec = _recorder;
        _recorder = null;
        if (rec == null) return;

        var done = new TaskCompletionSource();
        void Handler(object? _, RecordingCompleteEventArgs __) => done.TrySetResult();
        rec.OnRecordingComplete += Handler;
        try
        {
            rec.Stop();
            // Bounded wait: don't hang shutdown if the encoder stalls.
            await Task.WhenAny(done.Task, Task.Delay(TimeSpan.FromSeconds(5)));
        }
        catch { /* best effort */ }
        finally
        {
            try { rec.Dispose(); } catch { }
            _stream?.Dispose();
            _stream = null;
        }
        Console.WriteLine("✓ Screen video capture stopped");
    }

    private RecordingSourceBase ResolveSource(string id)
    {
        try
        {
            if (id.StartsWith("window:") && long.TryParse(id.AsSpan("window:".Length), out var handle))
            {
                var win = Recorder.GetWindows().FirstOrDefault(w => w.Handle == (IntPtr)handle);
                if (win != null) return new WindowRecordingSource(win.Handle);
                Console.Error.WriteLine($"⚠ video window '{id}' not found; recording the primary display");
            }
            else if (id.StartsWith("display:"))
            {
                var name = id.Substring("display:".Length);
                var disp = Recorder.GetDisplays().FirstOrDefault(d => d.DeviceName == name);
                if (disp != null) return new DisplayRecordingSource(disp.DeviceName);
                Console.Error.WriteLine($"⚠ video display '{id}' not found; recording the primary display");
            }
        }
        catch (Exception e)
        {
            Console.Error.WriteLine($"video source resolve failed ({e.Message}); using primary display");
        }
        return new DisplayRecordingSource(DisplayRecordingSource.MainMonitor);
    }

    private void OnChunk(byte[] data)
    {
        if (_stopped || data.Length == 0) return;
        if (!_firstFrameSeen)
        {
            _firstFrameSeen = true;
            FirstFrameDate = DateTime.UtcNow;
            OnFirstFrame?.Invoke();
        }
        _onSegment(data);
    }
}

/// <summary>
/// A write-only Stream that forwards every write to a callback (the video
/// socket send). ScreenRecorderLib writes fragmented-MP4 bytes here as they are
/// encoded, so nothing is staged to disk on the client. Reads/seeks are not
/// supported (the encoder only writes).
/// </summary>
internal sealed class CallbackStream : Stream
{
    private readonly Action<byte[]> _onWrite;
    private long _position;

    public CallbackStream(Action<byte[]> onWrite) { _onWrite = onWrite; }

    public override bool CanRead => false;
    public override bool CanSeek => false;
    public override bool CanWrite => true;
    public override long Length => _position;
    public override long Position { get => _position; set { } }

    public override void Write(byte[] buffer, int offset, int count)
    {
        if (count <= 0) return;
        var chunk = new byte[count];
        Buffer.BlockCopy(buffer, offset, chunk, 0, count);
        _position += count;
        _onWrite(chunk);
    }

    public override void Flush() { }
    public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
    public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
    public override void SetLength(long value) => throw new NotSupportedException();
}

File: lma-audio-capture-app-stack/source/macos/Sources/LMAAudioClient/VideoCapture.swift
import Foundation
import AVFoundation
import ScreenCaptureKit
import CoreMedia

/// Optional desktop-video capture: a SECOND ScreenCaptureKit stream (separate
/// from AudioCapture's) that captures the chosen display or window at a low
/// frame rate, H.264-encodes it via AVAssetWriter in fragmented-MP4 (CMAF)
/// mode, and hands each fMP4 segment to a callback for streaming.
///
/// Why a separate SCStream rather than adding .screen output to the audio
/// stream: the audio stream MUST keep its full-display filter (a window filter
/// would restrict captured system audio to that one app), while video should
/// capture exactly the display/window the user picked. Two streams keep the
/// two concerns independent — and video can fail/stop without touching audio.
///
/// Uses the SAME "Screen Recording" TCC permission the audio capture already
/// holds (ScreenCaptureKit requires it even for audio-only), so enabling video
/// adds no new permission prompt.
///
/// Encoder invariants the server/pipeline relies on:
///   • Fragmented MP4 (init segment first, then moof/mdat fragments) — the
///     server appends segments to a file that is playable/probe-able as it
///     grows, and ffmpeg muxes it with the call audio at call end.
///   • Keyframe at least every 2 seconds (segment interval forces this), so
///     future frame-extraction features can seek cheaply.
///   • Timestamps rebased to ZERO at the first captured frame — call-relative
///     alignment is carried separately as videoTimeOffsetMs on START_VIDEO.
@available(macOS 13.0, *)
final class VideoCapture: NSObject, SCStreamOutput, SCStreamDelegate, AVAssetWriterDelegate {

    /// A capturable video source (display or window) for the Settings picker.
    struct Source: Identifiable, Equatable {
        /// Stable-ish persisted id: "display:<displayID>" or "window:<windowID>".
        let id: String
        let name: String
    }

    /// Frames per second. Screen content changes slowly; 5 fps matches the
    /// Virtual Participant's recording and keeps CPU/bandwidth low.
    static let fps = 5
    /// Cap the longest edge; enough for legible slide text without huge bitrate.
    static let maxDimension = 1920
    /// Target H.264 bitrate. Screen content at 5 fps compresses very well.
    static let bitsPerSecond = 1_500_000

    private let onSegment: (Data) -> Void
    private let sourceID: String

    private var scStream: SCStream?
    private var writer: AVAssetWriter?
    private var writerInput: AVAssetWriterInput?
    private var sessionStarted = false
    private var firstFramePTS: CMTime = .invalid
    private(set) var firstFrameDate: Date?
    /// Called once when the first frame is captured (offset reporting).
    var onFirstFrame: (() -> Void)?

    private let encodeQueue = DispatchQueue(label: "lma.sck.video")
    private var stopped = false

    // Pause: the audio path CONSUMES AND DISCARDS samples while paused (its
    // recording is shorter than wall clock), so video must excise the same
    // span or everything after a pause is misaligned in the muxed MP4. While
    // paused we drop frames; on resume we subtract the accumulated pause time
    // from every subsequent PTS so the video timeline matches the audio's.
    private var paused = false
    private var pauseBegan: CMTime = .invalid
    private var pausedTotal: CMTime = .zero

    /// Mirror the audio mixer's pause state (called from the controller).
    /// The paused span itself is measured in capture-clock time by the frame
    /// handler: it opens at the first dropped frame and closes at the first
    /// post-resume frame.
    func setPaused(_ p: Bool) {
        encodeQueue.async { [self] in paused = p }
    }

    init(sourceID: String, onSegment: @escaping (Data) -> Void) {
        self.sourceID = sourceID
        self.onSegment = onSegment
        super.init()
    }

    // MARK: - Source enumeration (Settings picker)

    /// List capturable sources beyond the default: additional displays (the
    /// main display is the picker's built-in "Entire screen" = empty id), then
    /// on-screen windows that have a title (untitled/system chrome windows are
    /// noise, not meeting content).
    static func listSources() async -> [Source] {
        guard let content = try? await SCShareableContent.excludingDesktopWindows(
            true, onScreenWindowsOnly: true) else { return [] }
        var out: [Source] = []
        for (i, d) in content.displays.enumerated().dropFirst() {
            out.append(Source(id: "display:\(d.displayID)", name: "Display \(i + 1)"))
        }
        for w in content.windows {
            guard let title = w.title, !title.isEmpty,
                  let app = w.owningApplication, !app.applicationName.isEmpty,
                  w.frame.width >= 200, w.frame.height >= 150 // skip tiny utility windows
            else { continue }
            out.append(Source(id: "window:\(w.windowID)",
                              name: "\(app.applicationName) — \(title)"))
        }
        return out
    }

    // MARK: - Lifecycle

    func start() async throws {
        let content = try await SCShareableContent.excludingDesktopWindows(
            false, onScreenWindowsOnly: false)

        // Resolve the persisted source id; fall back to the main display when
        // it is gone (window closed, display unplugged, stale id after reboot).
        var filter: SCContentFilter
        var captureSize: CGSize
        if sourceID.hasPrefix("window:"),
           let wid = UInt32(sourceID.dropFirst("window:".count)),
           let win = content.windows.first(where: { $0.windowID == wid }) {
            filter = SCContentFilter(desktopIndependentWindow: win)
            captureSize = win.frame.size
        } else {
            var display = content.displays.first
            if sourceID.hasPrefix("display:"),
               let did = UInt32(sourceID.dropFirst("display:".count)),
               let d = content.displays.first(where: { $0.displayID == did }) {
                display = d
            } else if !sourceID.isEmpty && !sourceID.hasPrefix("display:") {
                print("⚠ video source '\(sourceID)' not found; recording the main display")
            }
            guard let d = display else {
                throw NSError(domain: "LMA", code: 2, userInfo: [
                    NSLocalizedDescriptionKey: "No display available for video capture."])
            }
            filter = SCContentFilter(display: d, excludingApplications: [], exceptingWindows: [])
            captureSize = CGSize(width: d.width, height: d.height)
        }

        // Scale to cap the longest edge; encoder wants even dimensions.
        let scale = min(1, CGFloat(Self.maxDimension) / max(captureSize.width, captureSize.height, 1))
        let w = max(2, Int(captureSize.width * scale) & ~1)
        let h = max(2, Int(captureSize.height * scale) & ~1)

        let cfg = SCStreamConfiguration()
        cfg.capturesAudio = false            // audio is AudioCapture's job
        cfg.width = w
        cfg.height = h
        cfg.minimumFrameInterval = CMTime(value: 1, timescale: CMTimeScale(Self.fps))
        cfg.queueDepth = 5
        cfg.showsCursor = true
        cfg.pixelFormat = kCVPixelFormatType_32BGRA

        try startWriter(width: w, height: h)

        let stream = SCStream(filter: filter, configuration: cfg, delegate: self)
        try stream.addStreamOutput(self, type: .screen, sampleHandlerQueue: encodeQueue)
        try await stream.startCapture()
        scStream = stream
        print("✓ Screen video capture started (\(w)x\(h) @ \(Self.fps) fps → fMP4)")
    }

    /// Stop capture and finish the writer, flushing the final segment(s).
    func stop() async {
        encodeQueue.sync { stopped = true }
        if let stream = scStream {
            try? await stream.stopCapture()
            scStream = nil
        }
        await withCheckedContinuation { (cont: CheckedContinuation<Void, Never>) in
            encodeQueue.async { [self] in
                guard let writer = writer, writer.status == .writing else {
                    cont.resume(); return
                }
                writerInput?.markAsFinished()
                writer.finishWriting {
                    cont.resume()
                }
            }
        }
        writer = nil
        writerInput = nil
        print("✓ Screen video capture stopped")
    }

    // MARK: - AVAssetWriter (fragmented MP4)

    private func startWriter(width: Int, height: Int) throws {
        // CMAF fragmented-MP4 profile: segments are delivered to the delegate
        // instead of a file. Every segment boundary is a keyframe, so the
        // 2-second interval is also our max keyframe interval.
        let writer = AVAssetWriter(contentType: .mpeg4Movie)
        writer.outputFileTypeProfile = .mpeg4CMAFCompliant
        writer.preferredOutputSegmentInterval = CMTime(seconds: 2, preferredTimescale: 1)
        writer.initialSegmentStartTime = .zero
        writer.delegate = self

        let settings: [String: Any] = [
            AVVideoCodecKey: AVVideoCodecType.h264,
            AVVideoWidthKey: width,
            AVVideoHeightKey: height,
            AVVideoCompressionPropertiesKey: [
                AVVideoAverageBitRateKey: Self.bitsPerSecond,
                AVVideoMaxKeyFrameIntervalKey: 2 * Self.fps,        // ≤2s in frames
                AVVideoMaxKeyFrameIntervalDurationKey: 2.0,          // ≤2s in time
                AVVideoProfileLevelKey: AVVideoProfileLevelH264MainAutoLevel,
                AVVideoAllowFrameReorderingKey: false,               // no B-frames
            ],
        ]
        let input = AVAssetWriterInput(mediaType: .video, outputSettings: settings)
        input.expectsMediaDataInRealTime = true
        guard writer.canAdd(input) else {
            throw NSError(domain: "LMA", code: 3, userInfo: [
                NSLocalizedDescriptionKey: "AVAssetWriter rejected the H.264 video input."])
        }
        writer.add(input)
        guard writer.startWriting() else {
            throw writer.error ?? NSError(domain: "LMA", code: 4, userInfo: [
                NSLocalizedDescriptionKey: "AVAssetWriter failed to start."])
        }
        self.writer = writer
        self.writerInput = input
    }

    /// fMP4 segments (init first, then media fragments) → the streaming callback.
    func assetWriter(_ writer: AVAssetWriter,
                     didOutputSegmentData segmentData: Data,
                     segmentType: AVAssetSegmentType,
                     segmentReport: AVAssetSegmentReport?) {
        onSegment(segmentData)
    }

    // MARK: - SCStreamOutput

    func stream(_ stream: SCStream, didOutputSampleBuffer sampleBuffer: CMSampleBuffer,
                of type: SCStreamOutputType) {
        guard type == .screen, !stopped,
              CMSampleBufferDataIsReady(sampleBuffer),
              let writer = writer, let input = writerInput else { return }

        // Skip idle/incomplete frames (SCK emits status attachments per frame).
        if let atts = CMSampleBufferGetSampleAttachmentsArray(sampleBuffer, createIfNecessary: false)
            as? [[SCStreamFrameInfo: Any]],
           let statusRaw = atts.first?[.status] as? Int,
           let status = SCFrameStatus(rawValue: statusRaw),
           status != .complete {
            return
        }

        let pts = CMSampleBufferGetPresentationTimeStamp(sampleBuffer)

        // Paused: drop the frame, but remember when the pause began (in
        // capture-clock time) so the span can be excised from the timeline.
        if paused {
            if !pauseBegan.isValid { pauseBegan = pts }
            return
        }
        if pauseBegan.isValid {
            // First frame after a resume: bank the paused span.
            pausedTotal = CMTimeAdd(pausedTotal, CMTimeSubtract(pts, pauseBegan))
            pauseBegan = .invalid
        }

        if !sessionStarted {
            guard writer.status == .writing else { return }
            firstFramePTS = pts
            firstFrameDate = Date()
            writer.startSession(atSourceTime: .zero)
            sessionStarted = true
            DispatchQueue.main.async { [weak self] in self?.onFirstFrame?() }
        }

        guard input.isReadyForMoreMediaData else { return } // drop frame under pressure

        // Rebase timestamps: media time starts at zero (call-relative
        // alignment is reported separately via videoTimeOffsetMs) and paused
        // spans are excised — matching the audio recording, which also skips
        // paused time.
        var timing = CMSampleTimingInfo(
            duration: CMTime(value: 1, timescale: CMTimeScale(Self.fps)),
            presentationTimeStamp: CMTimeSubtract(CMTimeSubtract(pts, firstFramePTS), pausedTotal),
            decodeTimeStamp: .invalid)
        var rebased: CMSampleBuffer?
        let status = CMSampleBufferCreateCopyWithNewTiming(
            allocator: kCFAllocatorDefault, sampleBuffer: sampleBuffer,
            sampleTimingEntryCount: 1, sampleTimingArray: &timing,
            sampleBufferOut: &rebased)
        guard status == noErr, let rb = rebased else { return }
        if !input.append(rb) {
            if writer.status == .failed {
                FileHandle.standardError.write(
                    "video encoder failed: \(writer.error.map { "\($0)" } ?? "unknown")\n".data(using: .utf8)!)
                stopped = true
            }
        }
    }

    func stream(_ stream: SCStream, didStopWithError error: Error) {
        FileHandle.standardError.write("SCK video stream stopped with error: \(error)\n".data(using: .utf8)!)
    }
}
